### PR TITLE
Add SDAR model support

### PR DESCRIPTION
**Pull request by chengshuang18**
## Motivation

Add **SDAR (`SDARForCausalLM`) model support** to SGLang so SDAR-style diffusion/LLM models can be served and evaluated via the existing SRT server + DLLM workflow (e.g., `--dllm-algorithm LowConfidence`). This enables running SDAR models with SGLang’s standard serving, CI, and evaluation tooling.

## Modifications

* Implement / integrate SDAR model support (`SDARForCausalLM`, `SDARMoeForCausalLM`) into the model registry / serving stack.
* Register SDAR-specific DLLM config in `dllm/config.py` (e.g., `block_size=4`, `mask_id=151669`).
* Add unit tests that cover:

  * Few-shot GSM8K accuracy + throughput sanity check.
  * Batch-size-1 generation speed sanity check and GitHub step summary output in CI.

## Accuracy Tests

Ran few-shot GSM8K evaluation (5-shot, 200 questions) on `SDAR-8B-Chat` with:

**Server args**

* `--trust-remote-code`
* `--mem-fraction-static 0.6`
* `--max-running-requests 128`
* `--attention-backend flashinfer`
* `--dllm-algorithm LowConfidence`
* `--tp 1`

**Eval args**

* `num_shots=5, num_questions=200, max_new_tokens=1024, parallel=128`

**Results**

* Accuracy: **0.940**
* Invalid: **0.000**
* Latency: **48.534 s** (200 questions)
* Output throughput: **659.163 token/s**

```
Accuracy: 0.940
Invalid: 0.000
Latency: 48.534 s
Output throughput: 659.163 token/s
metrics={'accuracy': 0.94, 'invalid': 0.0, 'latency': 48.53424892015755, 'output_throughput': 659.1633889839156}
```

Ran few-shot GSM8K evaluation (5-shot, 200 questions) on `SDAR-30B-A3B-Chat` with the same eval args:

**Results (TP1-EP1)**

* Accuracy: **0.930**
* Invalid: **0.000**
* Latency: **65.637 s** (200 questions)
* Output throughput: **467.036 token/s**

```
Accuracy: 0.930
Invalid: 0.000
Latency: 65.637 s
Output throughput: 467.036 token/s
[TP1-EP1] metrics={'accuracy': 0.93, 'invalid': 0.0, 'latency': 65.63730664178729, 'output_throughput': 467.0362263232145}
```

**Results (TP2-EP1)**

* Accuracy: **0.940**
* Invalid: **0.000**
* Latency: **59.353 s** (200 questions)
* Output throughput: **524.453 token/s**

```
Accuracy: 0.940
Invalid: 0.000
Latency: 59.353 s
Output throughput: 524.453 token/s
[TP2-EP1] metrics={'accuracy': 0.94, 'invalid': 0.0, 'latency': 59.353314684703946, 'output_throughput': 524.4525965459188}
```

**Results (TP2-EP2)**

* Accuracy: **0.950**
* Invalid: **0.000**
* Latency: **58.029 s** (200 questions)
* Output throughput: **528.671 token/s**

```
Accuracy: 0.950
Invalid: 0.000
Latency: 58.029 s
Output throughput: 528.671 token/s
[TP2-EP2] metrics={'accuracy': 0.95, 'invalid': 0.0, 'latency': 58.02852285839617, 'output_throughput': 528.6710481130434}
```

## Benchmarking and Profiling

Included a bs=1 speed sanity benchmark in unit tests (`test_bs_1_speed`) using:

* `max_new_tokens=2048`
* `tp=1`

Additionally, the GSM8K run reports **output throughput = 659.163 token/s** under the configuration above (see Accuracy Tests).

## Checklist

* [x] Format your code according to the [[Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit)](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
* [x] Add unit tests according to the [[Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests)](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
* [x] Update documentation according to [[Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations)](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
* [x] Provide accuracy and speed benchmark results according to [[Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy)](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [[Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed)](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
* [x] Follow the SGLang code style [[guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance)](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [[PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [[CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS)](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [[comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests)](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.

   * `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: test/registered/dllm/test_sdar.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    DEFAULT_URL_FOR_TEST,
+    CustomTestCase,
+    is_in_amd_ci,
+    is_in_ci,
+    popen_launch_server,
+    write_github_step_summary,
+)
+
+
+class TestLLaDA2Mini(CustomTestCase):
```
Comment (Monstertail) [inline test/registered/dllm/test_sdar.py]: Why class TestLLaDA2Mini here? Correct it to the right name.

File: python/sglang/srt/models/sdar.py [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+            allow_reduce_scatter=True,
+        )
+
+    def forward(
+        self,
+        positions: torch.Tensor,
+        hidden_states: torch.Tensor,
+        forward_batch: ForwardBatch,
+        residual: Optional[torch.Tensor],
+        post_residual_addition: Optional[torch.Tensor] = None,
```
Comment (Monstertail) [inline python/sglang/srt/models/sdar.py]: Could you please explain why we have post_residual_addition here? It looks like not being used.

File: python/sglang/srt/models/sdar.py [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+            hidden_states = hidden_states.bfloat16()
+
+        if not _is_npu or forward_batch.forward_mode.is_extend():
+            q, k, v = self.forward_prepare_native(
+                positions=positions,
+                hidden_states=hidden_states,
+                forward_batch=forward_batch,
+            )
+        else:
+            q, k, v = self.forward_prepare_npu(
```
Comment (ClawSeven) [inline python/sglang/srt/models/sdar.py]: Could you please share the accuracy and performance benchmarks if you have tested it on NPU hardware? If the NPU execution path is not validated, you may consider removing it.

File: python/sglang/srt/dllm/config.py [hunk 4]
```diff
@@ -35,6 +35,9 @@ def from_server_args(
         if model_config.hf_config.architectures[0] == "LLaDA2MoeModelLM":
             block_size = 32
             mask_id = 156895
+        elif model_config.hf_config.architectures[0] == "SDARForCausalLM":
```
Comment (ClawSeven) [inline python/sglang/srt/dllm/config.py]: Hi, could you take the refactoring advice from gemini-code-assist:
```
        DLLM_PARAMS = {
            "LLaDA2MoeModelLM": {"block_size": 32, "mask_id": 156895},
            "SDARForCausalLM": {"block_size": 4, "mask_id": 151669},
        }

        arch = model_config.hf_config.architectures[0]
        if arch in DLLM_PARAMS:
            params = DLLM_PARAMS[arch]
            block_size = params["block_size"]
            mask_id = params["mask_id"]
        else:
            raise RuntimeError(f"Unknown diffusion LLM: {arch}")
```

File: python/sglang/srt/models/sdar.py [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+        config: PretrainedConfig,
+        quant_config: Optional[QuantizationConfig] = None,
+        prefix: str = "",
+        alt_stream: Optional[torch.cuda.Stream] = None,
+    ):
+        super().__init__()
+        self.config = config
+        self.vocab_size = config.vocab_size
+        self.embed_dim = config.hidden_size
+        self.pp_group = get_pp_group()
```
Comment (ClawSeven) [inline python/sglang/srt/models/sdar.py:362]: Same question—I believe we haven't tested dLLM with pipeline parallelism. If it hasn't been tested, please remove the pipeline parallelism (PP) branch.

File: python/sglang/srt/models/sdar.py [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+
+import torch
+from torch import nn
+from transformers import PretrainedConfig
+
+from sglang.srt.distributed import get_pp_group, get_tensor_model_parallel_world_size
+from sglang.srt.layers.activation import SiluAndMul
+from sglang.srt.layers.communicator import LayerCommunicator, LayerScatterModes
+from sglang.srt.layers.dp_attention import (
+    get_attention_tp_rank,
```
Comment (ClawSeven) [inline python/sglang/srt/models/sdar.py:17]: By the way, I believe dLLM supports tensor parallelism (TP) and expert parallelism (EP). Could you please share the accuracy benchmarks with TP and EP enabled?

File: python/sglang/srt/models/sdar.py [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+        reduce_results: bool = True,
+        prefix: str = "",
+    ):
+        super().__init__()
+        self.tp_size = get_tensor_model_parallel_world_size()
+        self.gate_up_proj = MergedColumnParallelLinear(
+            config.hidden_size,
+            [config.intermediate_size] * 2,
+            bias=False,
+            quant_config=quant_config,
```
Comment (ClawSeven) [inline python/sglang/srt/models/sdar.py:66]: I believe the "tp_rank" and "tp_size" arguments should be parsing for MergedColumnParallelLinear for TP.

File: python/sglang/srt/models/sdar.py [hunk 8]
```diff
@@ -0,0 +1,616 @@
...
+            prefix=add_prefix("gate_up_proj", prefix),
+        )
+        self.down_proj = RowParallelLinear(
```
Comment (ClawSeven) [inline python/sglang/srt/models/sdar.py:69]: Same problem for "tp_rank" and "tp_size" arguments

File: python/sglang/srt/models/sdar.py [hunk 9]
```diff
@@ -0,0 +1,616 @@
...
+            config.hidden_size,
+            bias=False,
+            reduce_results=reduce_results,
+            quant_config=quant_config,
+            prefix=add_prefix("down_proj", prefix),
+        )
+        self.act_fn = SiluAndMul()
+
+    def forward(self, hidden_states: torch.Tensor, use_reduce_scatter: bool = False):
+        if (self.tp_size == 1) and hidden_states.shape[0] == 0:
```
Comment (ClawSeven) [inline python/sglang/srt/models/sdar.py]: Perhaps this check is unnecessary?

File: test/registered/dllm/test_sdar.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    def setUpClass(cls):
+        cls.model = "JetLM/SDAR-8B-Chat"
+        cls.base_url = DEFAULT_URL_FOR_TEST
+
+        other_args = [
+            "--trust-remote-code",
+            "--mem-fraction-static",
+            "0.6",
+            "--max-running-requests",
+            "128",
```
Comment (ClawSeven) [inline test/registered/dllm/test_sdar.py]: Perhaps a small batch size is better.

**Review by ClawSeven — CHANGES_REQUESTED** [review]

File: python/sglang/srt/models/sdar.py (hunk 7 above)
Comment (Monstertail) [reply python/sglang/srt/models/sdar.py:66]: @ClawSeven  I have checked this question before. Actually, in qwen 3, it does not have "tp_rank" and "tp_size" arguments because the https://github.com/sgl-project/sglang/blob/64950d8f9735e0e0f652092b695d21e3ca34f310/python/sglang/srt/models/qwen3.py#L102 will get those two parameters automatically if they are set None.

File: python/sglang/srt/models/sdar.py (hunk 8 above)
Comment (Monstertail) [reply python/sglang/srt/models/sdar.py:69]: The same case happens to RowParallelLinear https://github.com/sgl-project/sglang/blob/64950d8f9735e0e0f652092b695d21e3ca34f310/python/sglang/srt/models/qwen3.py#L113 , it will get those two arguments automatically like qwen 3, but I agree it is good to add those two and TP tests. @chengshuang18

File: python/sglang/srt/models/sdar.py (hunk 6 above)
Comment (Monstertail) [reply python/sglang/srt/models/sdar.py:17]: I thought SDAR only had dense models, so I ignored this. But I just found SDAR-30BA3B is one of the model series.  @chengshuang18 How do you support this MOE model? I think EP is necessary for MoE. In qwen3 series, qwen3_moe.py is in a separate file besides qwen3.py. We may also have a separate sdar_moe.py with EP.

File: python/sglang/srt/models/sdar.py [hunk 11]
```diff
@@ -139,8 +129,8 @@ def __init__(
             self.total_num_kv_heads,
             bias=getattr(config, "attention_bias", False),
             quant_config=quant_config,
-            tp_rank=attn_tp_rank,
-            tp_size=attn_tp_size,
+            # tp_rank=attn_tp_rank,
```
Comment (ClawSeven) [inline python/sglang/srt/models/sdar.py]: Remove the annotation

**Review by ClawSeven — APPROVED** [review]

**Comment by ClawSeven** [comment]
LGTM

File: test/registered/dllm/test_sdar_moe.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
+from sglang.test.ci.ci_register import register_amd_ci, register_cuda_ci
+
+register_cuda_ci(est_time=181, suite="stage-b-test-large-1-gpu")
```
Comment (ClawSeven) [inline test/registered/dllm/test_sdar_moe.py]: If setting tp4, "stage-b-test-large-1-gpu" is not suitable.

**Review by ClawSeven — CHANGES_REQUESTED** [review]

**Review by JustinTong0323 — APPROVED** [review]
LGTM
